debian/bongostat.1
.\" Man page generated from reStructuredText.
.
.TH "BONGOSTAT" "1" "January 30, 2015" "3.0" "bongodb-manual"
.SH NAME
bongostat \- BongoDB Use Statistics
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.SH SYNOPSIS
.sp
The \fBbongostat\fP utility provides a quick overview of the
status of a currently running \fBbongod\fP
or \fBbongos\fP
instance. \fBbongostat\fP is functionally similar to the
UNIX/Linux file system utility \fBvmstat\fP, but provides data regarding
\fBbongod\fP and \fBbongos\fP instances.
.sp
\fBSEE ALSO:\fP
.INDENT 0.0
.INDENT 3.5
For more information about monitoring BongoDB, see
http://docs.bongodb.org/manual/administration/monitoring\&.
.sp
For more background on various other BongoDB status outputs see:
.INDENT 0.0
.IP \(bu 2
http://docs.bongodb.org/manual/reference/command/serverStatus
.IP \(bu 2
http://docs.bongodb.org/manual/reference/command/replSetGetStatus
.IP \(bu 2
http://docs.bongodb.org/manual/reference/command/dbStats
.IP \(bu 2
http://docs.bongodb.org/manual/reference/command/collStats
.UNINDENT
.sp
For an additional utility that provides BongoDB metrics see
\fBbongotop\fP\&.
.UNINDENT
.UNINDENT
.SH ACCESS CONTROL REQUIREMENTS
.sp
In order to connect to a \fBbongod\fP that enforces authorization
with the \fI\-\-auth\fP option, specify the
\fI\-\-username\fP and \fI\-\-password\fP options, and the user specified must have the
\fBserverStatus\fP privilege action on the cluster resources.
.sp
The built\-in role \fBclusterMonitor\fP provides this privilege as
well as other privileges. To create a role with just the privilege to
run \fBbongostat\fP, see \fIcreate\-role\-for\-bongostat\fP\&.
.SH OPTIONS
.INDENT 0.0
.TP
.B bongostat
.UNINDENT
.INDENT 0.0
.TP
.B bongostat
.UNINDENT
.INDENT 0.0
.TP
.B \-\-help
Returns information on the options and use of \fBbongostat\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-verbose, \-v
Increases the amount of internal reporting returned on standard output
or in log files. Increase the verbosity with the \fB\-v\fP form by
including the option multiple times, (e.g. \fB\-vvvvv\fP\&.)
.UNINDENT
.INDENT 0.0
.TP
.B \-\-version
Returns the \fBbongostat\fP release number.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-host <hostname><:port>, \-h <hostname><:port>
\fIDefault\fP: localhost:27017
.sp
Specifies a resolvable hostname for the \fBbongod\fP to which to
connect. By default, the \fBbongostat\fP attempts to connect to a BongoDB
instance running on the localhost on port number \fB27017\fP\&.
.sp
To connect to a replica set, specify the
\fBreplSetName\fP and a seed list of set members, as in
the following:
.INDENT 7.0
.INDENT 3.5
.sp
.nf
.ft C
<replSetName>/<hostname1><:port>,<hostname2><:port>,<...>
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
You can always connect directly to a single BongoDB instance by
specifying the host and port number directly.
.sp
Changed in version 3.0.0: If you use IPv6 and use the \fB<address>:<port>\fP format, you must
enclose the portion of an address and port combination in
brackets (e.g. \fB[<address>]\fP).

.UNINDENT
.INDENT 0.0
.TP
.B \-\-port <port>
\fIDefault\fP: 27017
.sp
Specifies the TCP port on which the BongoDB instance listens for
client connections.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-ipv6
Enables IPv6 support and allows the \fBbongostat\fP to connect to the
BongoDB instance using an IPv6 network. All BongoDB programs and
processes disable IPv6 support by default.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-ssl
New in version 2.6.

.sp
Enables connection to a \fBbongod\fP or \fBbongos\fP that has
SSL support enabled.
.sp
The default distribution of BongoDB does not contain support for SSL.
For more information on BongoDB and SSL, see http://docs.bongodb.org/manual/tutorial/configure\-ssl\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-sslCAFile <filename>
New in version 2.6.

.sp
Specifies the \fB\&.pem\fP file that contains the root certificate chain
from the Certificate Authority. Specify the file name of the
\fB\&.pem\fP file using relative or absolute paths.
.sp
The default distribution of BongoDB does not contain support for SSL.
For more information on BongoDB and SSL, see http://docs.bongodb.org/manual/tutorial/configure\-ssl\&.
.sp
\fBWARNING:\fP
.INDENT 7.0
.INDENT 3.5
If the \fBbongo\fP shell or any other tool that connects to
\fBbongos\fP or \fBbongod\fP is run without
\fI\-\-sslCAFile\fP, it will not attempt to validate
server certificates. This results in vulnerability to expired
\fBbongod\fP and \fBbongos\fP certificates as well as to foreign
processes posing as valid \fBbongod\fP or \fBbongos\fP
instances. Ensure that you \fIalways\fP specify the CA file against which
server certificates should be validated in cases where intrusion is a
possibility.
.UNINDENT
.UNINDENT
.UNINDENT
.INDENT 0.0
.TP
.B \-\-sslPEMKeyFile <filename>
New in version 2.6.

.sp
Specifies the \fB\&.pem\fP file that contains both the SSL certificate
and key. Specify the file name of the \fB\&.pem\fP file using relative
or absolute paths.
.sp
This option is required when using the \fI\-\-ssl\fP option to connect
to a \fBbongod\fP or \fBbongos\fP that has
\fBCAFile\fP enabled \fIwithout\fP
\fBallowConnectionsWithoutCertificates\fP\&.
.sp
The default distribution of BongoDB does not contain support for SSL.
For more information on BongoDB and SSL, see http://docs.bongodb.org/manual/tutorial/configure\-ssl\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-sslPEMKeyPassword <value>
New in version 2.6.

.sp
Specifies the password to de\-crypt the certificate\-key file (i.e.
\fI\-\-sslPEMKeyFile\fP). Use the \fI\-\-sslPEMKeyPassword\fP option only if the
certificate\-key file is encrypted. In all cases, the \fBbongostat\fP will
redact the password from all logging and reporting output.
.sp
If the private key in the PEM file is encrypted and you do not specify
the \fI\-\-sslPEMKeyPassword\fP option, the \fBbongostat\fP will prompt for a passphrase. See
\fIssl\-certificate\-password\fP\&.
.sp
The default distribution of BongoDB does not contain support for SSL.
For more information on BongoDB and SSL, see http://docs.bongodb.org/manual/tutorial/configure\-ssl\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-sslCRLFile <filename>
New in version 2.6.

.sp
Specifies the \fB\&.pem\fP file that contains the Certificate Revocation
List. Specify the file name of the \fB\&.pem\fP file using relative or
absolute paths.
.sp
The default distribution of BongoDB does not contain support for SSL.
For more information on BongoDB and SSL, see http://docs.bongodb.org/manual/tutorial/configure\-ssl\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-sslAllowInvalidCertificates
New in version 2.6.

.sp
Bypasses the validation checks for server certificates and allows
the use of invalid certificates. When using the
\fBallowInvalidCertificates\fP setting, BongoDB logs as a
warning the use of the invalid certificate.
.sp
The default distribution of BongoDB does not contain support for SSL.
For more information on BongoDB and SSL, see http://docs.bongodb.org/manual/tutorial/configure\-ssl\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-sslAllowInvalidHostnames
New in version 3.0.

.sp
Disables the validation of the hostnames in SSL certificates. Allows
\fBbongostat\fP to connect to BongoDB instances if the hostname their
certificates do not match the specified hostname.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-sslFIPSMode
New in version 2.6.

.sp
Directs the \fBbongostat\fP to use the FIPS mode of the installed OpenSSL
library. Your system must have a FIPS compliant OpenSSL library to use
the \fI\-\-sslFIPSMode\fP option.
.sp
\fBNOTE:\fP
.INDENT 7.0
.INDENT 3.5
FIPS Compatible SSL is
available only in \fI\%BongoDB Enterprise\fP\&. See
http://docs.bongodb.org/manual/tutorial/configure\-fips for more information.
.UNINDENT
.UNINDENT
.UNINDENT
.INDENT 0.0
.TP
.B \-\-username <username>, \-u <username>
Specifies a username with which to authenticate to a BongoDB database
that uses authentication. Use in conjunction with the \fB\-\-password\fP and
\fB\-\-authenticationDatabase\fP options.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-password <password>, \-p <password>
Specifies a password with which to authenticate to a BongoDB database
that uses authentication. Use in conjunction with the \fB\-\-username\fP and
\fB\-\-authenticationDatabase\fP options.
.sp
If you do not specify an argument for \fI\-\-password\fP, \fBbongostat\fP will
prompt interactively for a password on the console.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-authenticationDatabase <dbname>
New in version 2.4.

.sp
Specifies the database that holds the user\(aqs credentials.
.sp
\fI\-\-authenticationDatabase\fP is required for \fBbongod\fP
and \fBbongos\fP instances that use \fIauthentication\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-authenticationMechanism <name>
\fIDefault\fP: BONGODB\-CR
.sp
New in version 2.4.

.sp
Changed in version 2.6: Added support for the \fBPLAIN\fP and \fBBONGODB\-X509\fP authentication
mechanisms.

.sp
Specifies the authentication mechanism the \fBbongostat\fP instance uses to
authenticate to the \fBbongod\fP or \fBbongos\fP\&.
.TS
center;
|l|l|.
_
T{
Value
T}	T{
Description
T}
_
T{
BONGODB\-CR
T}	T{
BongoDB challenge/response authentication.
T}
_
T{
BONGODB\-X509
T}	T{
BongoDB SSL certificate authentication.
T}
_
T{
PLAIN
T}	T{
External authentication using LDAP. You can also use \fBPLAIN\fP
for authenticating in\-database users. \fBPLAIN\fP transmits
passwords in plain text. This mechanism is available only in
\fI\%BongoDB Enterprise\fP\&.
T}
_
T{
GSSAPI
T}	T{
External authentication using Kerberos. This mechanism is
available only in \fI\%BongoDB Enterprise\fP\&.
T}
_
.TE
.UNINDENT
.INDENT 0.0
.TP
.B \-\-gssapiServiceName
New in version 2.6.

.sp
Specify the name of the service using \fBGSSAPI/Kerberos\fP\&. Only required if the service does not use the
default name of \fBbongodb\fP\&.
.sp
This option is available only in BongoDB Enterprise.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-gssapiHostName
New in version 2.6.

.sp
Specify the hostname of a service using \fBGSSAPI/Kerberos\fP\&. \fIOnly\fP required if the hostname of a machine does
not match the hostname resolved by DNS.
.sp
This option is available only in BongoDB Enterprise.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-noheaders
Disables the output of column or field names.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-rowcount <number>, \-n <number>
Controls the number of rows to output. Use in conjunction with
the \fBsleeptime\fP argument to control the duration of a
\fBbongostat\fP operation.
.sp
Unless \fI\-\-rowcount\fP is specified, \fBbongostat\fP
will return an infinite number of rows (e.g. value of \fB0\fP\&.)
.UNINDENT
.INDENT 0.0
.TP
.B \-\-http
Configures \fBbongostat\fP to collect data using the HTTP interface
rather than a raw database connection.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-discover
Discovers and reports on statistics from all members of a \fIreplica
set\fP or \fIsharded cluster\fP\&. When connected to any member of a
replica set, \fI\%\-\-discover\fP all non\-\fIhidden members\fP of the replica set. When connected to a \fBbongos\fP,
\fBbongostat\fP will return data from all \fIshards\fP in
the cluster. If a replica set provides a shard in the sharded cluster,
\fBbongostat\fP will report on non\-hidden members of that replica
set.
.sp
The \fIbongostat \-\-host\fP option is not required but
potentially useful in this case.
.sp
Changed in version 2.6: When running with \fI\%\-\-discover\fP, \fBbongostat\fP now
respects :option:\-\-rowcount\(ga.

.UNINDENT
.INDENT 0.0
.TP
.B \-\-all
Configures \fBbongostat\fP to return all optional \fI\%fields\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-json
New in version 3.0.0.

.sp
Returns output for \fBbongostat\fP in \fIJSON\fP format.
.UNINDENT
.INDENT 0.0
.TP
.B <sleeptime>
The final argument is the length of time, in seconds, that
\fBbongostat\fP waits in between calls. By default \fBbongostat\fP
returns one call every second.
.sp
\fBbongostat\fP returns values that reflect the operations
over a 1 second period. For values of \fB<sleeptime>\fP greater
than 1, \fBbongostat\fP averages data to reflect average
operations per second.
.UNINDENT
.SH FIELDS
.sp
\fBbongostat\fP returns values that reflect the operations over a
1 second period. When \fBbongostat <sleeptime>\fP has a value
greater than 1, \fBbongostat\fP averages the statistics to reflect
average operations per second.
.sp
\fBbongostat\fP outputs the following fields:
.INDENT 0.0
.TP
.B inserts
The number of objects inserted into the database per second. If
followed by an asterisk (e.g. \fB*\fP), the datum refers to a
replicated operation.
.UNINDENT
.INDENT 0.0
.TP
.B query
The number of query operations per second.
.UNINDENT
.INDENT 0.0
.TP
.B update
The number of update operations per second.
.UNINDENT
.INDENT 0.0
.TP
.B delete
The number of delete operations per second.
.UNINDENT
.INDENT 0.0
.TP
.B getmore
The number of get more (i.e. cursor batch) operations per second.
.UNINDENT
.INDENT 0.0
.TP
.B command
The number of commands per second. On \fIslave\fP and
\fIsecondary\fP systems, \fBbongostat\fP presents two values
separated by a pipe character (e.g. \fB|\fP), in the form of
\fBlocal|replicated\fP commands.
.UNINDENT
.INDENT 0.0
.TP
.B flushes
The number of \fIfsync\fP operations per second.
.UNINDENT
.INDENT 0.0
.TP
.B mapped
The total amount of data mapped in megabytes. This is the total
data size at the time of the last \fBbongostat\fP call.
.UNINDENT
.INDENT 0.0
.TP
.B size
The amount of virtual memory in megabytes used by the process at
the time of the last \fBbongostat\fP call.
.UNINDENT
.INDENT 0.0
.TP
.B non\-mapped
The total amount of virtual memory excluding all mapped memory at
the time of the last \fBbongostat\fP call.
.UNINDENT
.INDENT 0.0
.TP
.B res
The amount of resident memory in megabytes used by the process at
the time of the last \fBbongostat\fP call.
.UNINDENT
.INDENT 0.0
.TP
.B faults
Changed in version 2.1.

.sp
The number of page faults per second.
.sp
Before version 2.1 this value was only provided for BongoDB
instances running on Linux hosts.
.UNINDENT
.INDENT 0.0
.TP
.B locked
The percent of time in a global write lock.
.sp
Changed in version 2.2: The \fBlocked db\fP field replaces the \fBlocked %\fP field to
more appropriate data regarding the database specific locks in
version 2.2.

.UNINDENT
.INDENT 0.0
.TP
.B locked db
New in version 2.2.

.sp
The percent of time in the per\-database context\-specific
lock. \fBbongostat\fP will report the database that has spent
the most time since the last \fBbongostat\fP call with a write
lock.
.sp
This value represents the amount of time that the listed database
spent in a locked state \fIcombined\fP with the time that the
\fBbongod\fP spent in the global lock. Because of this, and
the sampling method, you may see some values greater than 100%.
.UNINDENT
.INDENT 0.0
.TP
.B idx miss
The percent of index access attempts that required a page fault
to load a btree node. This is a sampled value.
.UNINDENT
.INDENT 0.0
.TP
.B qr
The length of the queue of clients waiting to read data from the
BongoDB instance.
.UNINDENT
.INDENT 0.0
.TP
.B qw
The length of the queue of clients waiting to write data from the
BongoDB instance.
.UNINDENT
.INDENT 0.0
.TP
.B ar
The number of active clients performing read operations.
.UNINDENT
.INDENT 0.0
.TP
.B aw
The number of active clients performing write operations.
.UNINDENT
.INDENT 0.0
.TP
.B netIn
The amount of network traffic, in \fIbytes\fP, received by the BongoDB instance.
.sp
This includes traffic from \fBbongostat\fP itself.
.UNINDENT
.INDENT 0.0
.TP
.B netOut
The amount of network traffic, in \fIbytes\fP, sent by the BongoDB instance.
.sp
This includes traffic from \fBbongostat\fP itself.
.UNINDENT
.INDENT 0.0
.TP
.B conn
The total number of open connections.
.UNINDENT
.INDENT 0.0
.TP
.B set
The name, if applicable, of the replica set.
.UNINDENT
.INDENT 0.0
.TP
.B repl
The replication status of the member.
.TS
center;
|l|l|.
_
T{
\fBValue\fP
T}	T{
\fBReplication Type\fP
T}
_
T{
M
T}	T{
\fImaster\fP
T}
_
T{
SEC
T}	T{
\fIsecondary\fP
T}
_
T{
REC
T}	T{
recovering
T}
_
T{
UNK
T}	T{
unknown
T}
_
T{
SLV
T}	T{
\fIslave\fP
T}
_
T{
RTR
T}	T{
bongos process ("router")
T}
_
.TE
.UNINDENT
.SH USAGE
.sp
In the first example, \fBbongostat\fP will return data every
second for 20 seconds. \fBbongostat\fP collects data from the
\fBbongod\fP instance running on the localhost interface on
port 27017. All of the following invocations produce identical
behavior:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
bongostat \-\-rowcount 20 1
bongostat \-\-rowcount 20
bongostat \-n 20 1
bongostat \-n 20
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
In the next example, \fBbongostat\fP returns data every 5 minutes
(or 300 seconds) for as long as the program runs. \fBbongostat\fP
collects data from the \fBbongod\fP instance running on the
localhost interface on port \fB27017\fP\&. Both of the following
invocations produce identical behavior.
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
bongostat \-\-rowcount 0 300
bongostat \-n 0 300
bongostat 300
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
In the following example, \fBbongostat\fP returns data every 5
minutes for an hour (12 times.) \fBbongostat\fP collects data
from the \fBbongod\fP instance running on the localhost interface
on port 27017. Both of the following invocations produce identical
behavior.
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
bongostat \-\-rowcount 12 300
bongostat \-n 12 300
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
In many cases, using the \fI\-\-discover\fP
will help provide a more complete snapshot of the state of an entire
group of machines. If a \fBbongos\fP process connected to a
\fIsharded cluster\fP is running on port \fB27017\fP of the local
machine, you can use the following form to return statistics from all
members of the cluster:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
bongostat \-\-discover
.ft P
.fi
.UNINDENT
.UNINDENT
.SH AUTHOR
BongoDB Documentation Project
.SH COPYRIGHT
2011-2015
.\" Generated by docutils manpage writer.
.


NEWS
WiredTiger release 2.9.0, 2016-09-06
------------------------------------

New features and API changes; refer to the API documentation for full details:

* WT-2360	Add nesting and disjunction functionality to cursor joins
* WT-2552	Add a public API to allow custom filesystem implementations
* WT-2738	Remove the ability to change the default checkpoint name. See the upgrading documentation for details.
* WT-2711	Change the statistics log configuration options. Restrict where statistics log files can be created to remove security vulnerability. See upgrading documentation for details.
* WT-2764	Enhance the checkpoint implementation, which led to a change in default eviction settings. See the upgrading documentation for details.
* WT-2880	Add Zstandard compression support
* See the upgrading documentation for information about renamed and removed statistics.

Significant changes and bug fixes:

* WT-2026	Fix a bug where in-memory pages were allowed to grow too large
* WT-2343	Ensure we cannot rename/drop when a backup cursor is open
* WT-2520	Ensure WT_SESSION::verify does not alter tables
* WT-2576	Ensure variable-length column-store cannot out-of-order return
* WT-2664	Change eviction so any eviction thread can find candidates
* WT-2702	Resolve an issue where under high thread load, WiredTiger exceeds cache size
* WT-2731	Resolve an issue where raw compression can create pages that are larger than expected
* WT-2737	Allow reconciliation to scrub dirty pages rather than evicting them
* WT-2757	Fix a bug in column stores where it was not possible to retrieve the new record number after an append
* WT-2816	Improve eviction performance

Issues fixed in MongoDB:

* SERVER-23504	Resolve a resource leak
* SERVER-23588	Stop using _open_osfhandle on Windows
* SERVER-23659	Improve log error message at startup
* SERVER-23661	Fix an issue where $sample takes disproportionately long time on newly created collection
* SERVER-24306	Fix stall in log_flush switching to new files
* SERVER-24580	Add more eviction stats to track efficiency
* SERVER-24580	Enhance eviction when application threads are contributing
* SERVER-24971	Don't cache buffers after application eviction
* SERVER-25843	Remove a redundant test
* SERVER-25845	Resolve an explicit null dereferenced
* SERVER-25846	Resolve a dereference after null with a check
* SERVER-26753	Don't spin on a read-lock in a tight loop

Other noteworthy changes since the previous release:

* WT-2103	Add incremental backup testing to test/format
* WT-2330	Ensure In-memory configurations do not create on-disk collection files
* WT-2450	Resolve an issue where salvage releases pages, then explicitly evicts them. Which can race with the eviction server
* WT-2453	Add multiple eviction queues
* WT-2504	Remove READONLY conditional for base config
* WT-2505	Resolve clang analyzer warnings
* WT-2508	Ensure test programs should remove test directories on the "clean" target
* WT-2512	Change how wtperf throttle values per thread are calculated
* WT-2513	Fix a type conversion warning
* WT-2517	Resolve incorrect uses of setvbuf on Windows
* WT-2518	Optimization of LSM checkpoint handle acquisition
* WT-2522	Resolve incorrect format code in message
* WT-2525	Miscellaneous cleanups of In-memory configurations
* WT-2526	Prevent mixing and matching readonly file-handles and read/write data-handles
* WT-2528	Fix a style error in WiredTiger build
* WT-2529	Be less aggressive asserting in readonly connections
* WT-2531	Resolve a case where in-memory tables are wasting space in truncation
* WT-2532	Change the WT_STREAM_APPEND and WT_STREAM_LINE_BUFFER flags as they overlap
* WT-2533	Don't let in-memory tables return a zero size
* WT-2534	Use atomic add when allocating transaction IDs
* WT-2535	Extend test/format to test for transactions reading their writes
* WT-2537	Check for an old and new uninitialized LSN for recovery
* WT-2539	Make streams a separate handle type from files
* WT-2540	Separate stream and file handle methods
* WT-2542	Resolve fixed-length column store reconciliation overwriting original values
* WT-2544	Fix eviction stats when clear is used
* WT-2546	Change evict server to be smarter when it decides whether or not to evict pages
* WT-2547	Add eviction testing configurations for 1 thread
* WT-2548	Cap the amount of data handed to raw compression
* WT-2550	Add support for raw mode cursors in java
* WT-2553	Add some In-memory documentation and stress testing
* WT-2556	Fix a typo in the Java example code
* WT-2557	Ensure that we truncate logs before closing backup cursor
* WT-2558	Reorganize WT_PAGE structure
* WT-2559	Resolve a potential segfault in Windows logging code
* WT-2560	Add a rwlock to protect transaction state, don't spin on this lock
* WT-2562	Add sleep loop to adapt for slow machines like PPC
* WT-2565	Further limit the amount of data passed to raw compression
* WT-2566	Add explicit memory barriers to Lock/unlock operations
* WT-2567	Add a check to see if logging is enabled for truncate rather than assuming it is
* WT-2568	Add a backward compatible constructor for PackInputStream
* WT-2569	Ensure __win_handle_read is followed by a call to GetLastError()
* WT-2570	Minor lint and cleanups to code
* WT-2571	Join code clean up
* WT-2572	Ensure test/format doesn't select an incompatible set of options for in-memory options configured
* WT-2573	Resolve a potential case where a double free can occur
* WT-2574	Ensure test/format frees all allocated configure memory
* WT-2576	Ensure variable-length column-store cannot out-of-order return
* WT-2577	Resolve a potential double-free following an in-memory split
* WT-2579	Change so diagnostic only code can use file streams directly
* WT-2580	Fix potential SWIG naming conflict in Java
* WT-2581	Resolve a case where we fail to track saved updates during a page split
* WT-2582	Fix a case where eviction can return WT_RESTART when running in-memory
* WT-2584	Change to not use periods at the end of error messages

src/bongo/gotools/bongostat/main/bongostat.go
// Main package for the bongostat tool.
package main

import (
	"os"
	"strconv"
	"strings"
	"time"

	"github.com/bongodb/bongo-tools/common/log"
	"github.com/bongodb/bongo-tools/common/options"
	"github.com/bongodb/bongo-tools/common/password"
	"github.com/bongodb/bongo-tools/common/signals"
	"github.com/bongodb/bongo-tools/common/util"
	"github.com/bongodb/bongo-tools/bongostat"
	"github.com/bongodb/bongo-tools/bongostat/stat_consumer"
	"github.com/bongodb/bongo-tools/bongostat/stat_consumer/line"
	"github.com/bongodb/bongo-tools/bongostat/status"
)

// optionKeyNames interprets the CLI options Columns and AppendColumns into
// the internal keyName mapping.
func optionKeyNames(option string) map[string]string {
	kn := make(map[string]string)
	columns := strings.Split(option, ",")
	for _, column := range columns {
		naming := strings.Split(column, "=")
		if len(naming) == 1 {
			kn[naming[0]] = naming[0]
		} else {
			kn[naming[0]] = naming[1]
		}
	}
	return kn
}

// optionCustomHeaders interprets the CLI options Columns and AppendColumns
// into a list of custom headers.
func optionCustomHeaders(option string) (headers []string) {
	columns := strings.Split(option, ",")
	for _, column := range columns {
		naming := strings.Split(column, "=")
		headers = append(headers, naming[0])
	}
	return
}

func main() {
	// initialize command-line opts
	opts := options.New(
		"bongostat",
		bongostat.Usage,
		options.EnabledOptions{Connection: true, Auth: true, Namespace: false})
	opts.UseReadOnlyHostDescription()

	// add bongostat-specific options
	statOpts := &bongostat.StatOptions{}
	opts.AddOptions(statOpts)

	interactiveOption := opts.FindOptionByLongName("interactive")
	if _, available := stat_consumer.FormatterConstructors["interactive"]; !available {
		// make --interactive inaccessible
		interactiveOption.LongName = ""
		interactiveOption.ShortName = 0
	}

	args, err := opts.Parse()
	if err != nil {
		log.Logvf(log.Always, "error parsing command line options: %v", err)
		log.Logvf(log.Always, "try 'bongostat --help' for more information")
		os.Exit(util.ExitBadOptions)
	}

	log.SetVerbosity(opts.Verbosity)
	signals.Handle()

	sleepInterval := 1
	if len(args) > 0 {
		if len(args) != 1 {
			log.Logvf(log.Always, "too many positional arguments: %v", args)
			log.Logvf(log.Always, "try 'bongostat --help' for more information")
			os.Exit(util.ExitBadOptions)
		}
		sleepInterval, err = strconv.Atoi(args[0])
		if err != nil {
			log.Logvf(log.Always, "invalid sleep interval: %v", args[0])
			os.Exit(util.ExitBadOptions)
		}
		if sleepInterval < 1 {
			log.Logvf(log.Always, "sleep interval must be at least 1 second")
			os.Exit(util.ExitBadOptions)
		}
	}

	// print help, if specified
	if opts.PrintHelp(false) {
		return
	}

	// print version, if specified
	if opts.PrintVersion() {
		return
	}

	if opts.Auth.Username != "" && opts.Auth.Source == "" && !opts.Auth.RequiresExternalDB() {
		log.Logvf(log.Always, "--authenticationDatabase is required when authenticating against a non $external database")
		os.Exit(util.ExitBadOptions)
	}

	if statOpts.Interactive && statOpts.Json {
		log.Logvf(log.Always, "cannot use output formats --json and --interactive together")
		os.Exit(util.ExitBadOptions)
	}

	if statOpts.Deprecated && !statOpts.Json {
		log.Logvf(log.Always, "--useDeprecatedJsonKeys can only be used when --json is also specified")
		os.Exit(util.ExitBadOptions)
	}

	if statOpts.Columns != "" && statOpts.AppendColumns != "" {
		log.Logvf(log.Always, "-O cannot be used if -o is also specified")
		os.Exit(util.ExitBadOptions)
	}

	if statOpts.HumanReadable != "true" && statOpts.HumanReadable != "false" {
		log.Logvf(log.Always, "--humanReadable must be set to either 'true' or 'false'")
		os.Exit(util.ExitBadOptions)
	}

	// we have to check this here, otherwise the user will be prompted
	// for a password for each discovered node
	if opts.Auth.ShouldAskForPassword() {
		opts.Auth.Password = password.Prompt()
	}

	var factory stat_consumer.FormatterConstructor
	if statOpts.Json {
		factory = stat_consumer.FormatterConstructors["json"]
	} else if statOpts.Interactive {
		factory = stat_consumer.FormatterConstructors["interactive"]
	} else {
		factory = stat_consumer.FormatterConstructors[""]
	}
	formatter := factory(statOpts.RowCount, !statOpts.NoHeaders)

	cliFlags := 0
	if statOpts.Columns == "" {
		cliFlags = line.FlagAlways
		if statOpts.Discover {
			cliFlags |= line.FlagDiscover
			cliFlags |= line.FlagHosts
		}
		if statOpts.All {
			cliFlags |= line.FlagAll
		}
		if strings.Contains(opts.Host, ",") {
			cliFlags |= line.FlagHosts
		}
	}

	var customHeaders []string
	if statOpts.Columns != "" {
		customHeaders = optionCustomHeaders(statOpts.Columns)
	} else if statOpts.AppendColumns != "" {
		customHeaders = optionCustomHeaders(statOpts.AppendColumns)
	}

	var keyNames map[string]string
	if statOpts.Deprecated {
		keyNames = line.DeprecatedKeyMap()
	} else if statOpts.Columns == "" {
		keyNames = line.DefaultKeyMap()
	} else {
		keyNames = optionKeyNames(statOpts.Columns)
	}
	if statOpts.AppendColumns != "" {
		addKN := optionKeyNames(statOpts.AppendColumns)
		for k, v := range addKN {
			keyNames[k] = v
		}
	}

	readerConfig := &status.ReaderConfig{
		HumanReadable: statOpts.HumanReadable == "true",
	}
	if statOpts.Json {
		readerConfig.TimeFormat = "15:04:05"
	}

	consumer := stat_consumer.NewStatConsumer(cliFlags, customHeaders,
		keyNames, readerConfig, formatter, os.Stdout)
	seedHosts := util.CreateConnectionAddrs(opts.Host, opts.Port)
	var cluster bongostat.ClusterMonitor
	if statOpts.Discover || len(seedHosts) > 1 {
		cluster = &bongostat.AsyncClusterMonitor{
			ReportChan:    make(chan *status.ServerStatus),
			ErrorChan:     make(chan *status.NodeError),
			LastStatLines: map[string]*line.StatLine{},
			Consumer:      consumer,
		}
	} else {
		cluster = &bongostat.SyncClusterMonitor{
			ReportChan: make(chan *status.ServerStatus),
			ErrorChan:  make(chan *status.NodeError),
			Consumer:   consumer,
		}
	}

	var discoverChan chan string
	if statOpts.Discover {
		discoverChan = make(chan string, 128)
	}

	opts.Direct = true
	_, setName := util.ParseConnectionString(opts.Host)
	opts.ReplicaSetName = setName
	stat := &bongostat.BongoStat{
		Options:       opts,
		StatOptions:   statOpts,
		Nodes:         map[string]*bongostat.NodeMonitor{},
		Discovered:    discoverChan,
		SleepInterval: time.Duration(sleepInterval) * time.Second,
		Cluster:       cluster,
	}

	for _, v := range seedHosts {
		stat.AddNewNode(v)
	}

	// kick it off
	err = stat.Run()
	if err != nil {
		log.Logvf(log.Always, "Failed: %v", err)
		os.Exit(util.ExitError)
	}
}
